### ISSUE TITLE
STS considered harmful - shouldn't be in 3.3 core spec

### ISSUE THREAD

**Issue by kaniini** [ircv3/ircv3-specifications#207]
Not to be that guy, but I question why STS should be a core component.

STS as a method to advertise 'secure-preferred' is harmful.  Security as a *preference* implies that a connection *without* security is also somehow accepted.  In practice, this means that conversations can still be compromised by an unencrypted user, because very few people use `+S` and equivalent channel modes.

The real solution for secure IRC usage, must be TLS from the first byte, as the STS implementation leaks metadata (specifically, if you `/server irc.indymedia.org` for example, an attacker doing passive sniffing now has metadata that you attempted to connect to `irc.indymedia.org`).  Using DNS SRV records would actually be an appropriate solution here.

Mandating STS as a solution to the IRC security problem is half-baked at best, and worse, encourages the status quo to continue (since there is a downgrade path).  STS is also ineffective in the IRC environment without dependence on CAs or fingerprinting, which should be included in the STS declaration (but the spec does not really mention this).

I suspect that adding things to the core like STS will result in many vendors simply ignoring that component of the spec, thus diluting the meaning between "core" and "extension".

A better solution would be to publish a "best operational practices" document which includes this and other advice about operating secure IRC networks.  If that is acceptable, I will be happy to contribute a draft for one.

**Comment by SaberUK** [comment]
:-1:  The fact is that whilst TLS only should be the future you will not get people to upgrade by just hard dropping TLS. Whilst STS has its flaws its considerably better than `STARTTLS` for upgrading plaintext connections to TLS. If you can provide a better option then feel free to state it.

**Comment by ChickenNuggers** [comment]
:-1: Baby steps, not giant leaps. I agree with @SaberUK that while TLS is better than STS, STS is better than `STARTTLS`; we should at least try moving to that first without forcing TLS.

**Comment by kaniini** [comment]
Note that I am totally fine with it being an *extension*.  Having it as *core* under the current policy means that it must be enabled for an IRC deployment to be considered *compliant*.

Under no circumstance is your typical IRC network ready to deploy this.

**Comment by jwheare** [comment]
How do we encourage server and client developers to build this in (behind a server config flag) without it being core? The point is that server operators need to be given the ability to enable this for IRC to move towards being more secure. They can't do that if the IRCd they rely on decided it was optional to include.

**Comment by kaniini** [comment]
Don't know, but it needs to be addressed.  I will *not* merge support for this until such time that it is addressed.

As I said earlier, "under no circumstance is your typical IRC network ready to deploy this".

**Comment by attilamolnar** [comment]
You have a misunderstanding. Clients receiving an STS policy MUST follow it. It is not "secure preferred". It's "from now on if you are sts compliant, you MUST use tls for future connections to this server." There is no choice for compliant implementations, they must switch to tls and keep using it in the future.

The consensus is that deprecating plaintext in one go is not realistic and if any such spec is written, clients will not follow it simply because they want to be able to connect to Quakenet etc.

STARTTLS is a protocol mechanism not a policy and thus it is trying to achieve a completely different goal.

**Comment by kaniini** [comment]
Okay, well, either:

* Give network *operators* an *explicit opt-out*, or
* Move it to *extensions*.

Otherwise it will never ever happen in charybdis.

**Comment by ChickenNuggers** [comment]
Are you saying that network operators should be able to opt out of an enforced encrypted connection?..

**Comment by jwheare** [comment]
A config option for operators is absolutely necessary, and obviously so. They get to decide whether to implement it or not, whether to even use transport security at all, which certs to use etc. It should absolutely be optional for them.

It should NOT be optional for client/server developers that want to continue supporting IRCv3 specifications.

**Comment by SaberUK** [comment]
>Give network operators an explicit opt-out, or

IRCv3 has never required that *networks* have features, only *implementations*. Thats why we have a software list on the website not a network list.

**Comment by dequis** [comment]
How about rewording this part of the spec?

>The capability value, which MUST always be present

It sounds like it's saying it's required for servers to always advertise the sts cap. And we don't want to force this on all server admins

Maybe it's a matter of adding a conditional like "if the server is configured to use a STS policy"

**Comment by attilamolnar** [comment]
@kaniini If your worry is that server operators cannot opt out then that's not true. Servers can advertise any sts policy they like and that includes a no enforcement policy.

**Comment by ChickenNuggers** [comment]
>How about rewording this part of the spec?

>>The capability value, which MUST always be present
>It sounds like it's saying it's required for servers to always advertise the sts cap. And we don't want >to force this on all server admins
>
>Maybe it's a matter of adding a conditional like "if the server is configured to use a STS policy"

Say something along the lines of "MUST always be present **after advertised**"

**Comment by kaniini** [comment]
It should be *MAY* or "MUST if an STS policy is specified by the administrator".  This must be explicit.

**Comment by Aerdan** [comment]
I think STS should be dropped entirely, and replaced with an informative document requesting that clients default to [SSL on port 6697](https://tools.ietf.org/html/rfc7194) and explicitly require user intervention to connect to plaintext IRC. 

Both STS and `CAP tls` have the problem that they can be stripped before the client can see them, thereby creating a downgrade attack. Moreover, because you're connecting plaintext-first, you're still subject to all the problems that plaintext IRC has.

**Comment by ChickenNuggers** [comment]
@Aerdan I'm drafting something up for that already. :D

**Comment by SoniEx2** [comment]
Isn't STS supposed to be like [HSTS](https://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security)? Basically the browser/IRC client should track the status and stuff for a given time? (Read the linked wikipedia page)

**Comment by ChickenNuggers** [comment]
There's an existing document you can read about STS.

**Comment by SoniEx2** [comment]
Oh it doesn't specify a default port. It also doesn't allow (multiple) alternative ports. Yes that's an issue.
